src/Microsoft.Build.UnGAC/Program.cs

```diff
@@ -0,0 +1,114 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.IO;
+using System.Linq;
+using System.Runtime.ExceptionServices;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace Microsoft.Build.UnGAC
+{
+    class Program
+    {
+        static void Main(string[] args)
+        {
+            try
+            {
+                List<string> allInstancesOfGACUtil = new List<string>();
+                StreamReader output;
+                StreamWriter input;
+
+                Process proc = new Process
```

Comment:
This seems heavy. Is there a way to run it in-proc?

Reply:
I'm not aware of one. I considered using an `exec` task but we need this to be an exe. If someone knows I'd love to know!

Reply:
@ladipro?
He knows everything.

Reply:
You can access the GAC via the Fusion interface, for example:
https://docs.microsoft.com/en-us/dotnet/framework/unmanaged-api/fusion/iassemblycache-uninstallassembly-method

Not sure if a COM interop wrapper exists or you'll have to build one yourself. I guess a big reason to go this route would be if gacutil.exe is not guaranteed to be present on the machine. Otherwise, if you know for sure there is gacutil.exe and you can find it, it may be easier to just execute it. I'm assuming we don't care about performance here.

Reply:
Good idea. There are a few that exist, but either the links to them are dead or they're not supported anymore. I found https://github.com/EasyHook/EasyHook
http://blog.wibeck.org/2008/10/working-with-gac-from-code/ points to [an API wrapper](https://docs.microsoft.com/en-us/archive/blogs/junfeng/articles/229649) that unfortunately 404's today.
There's also https://www.codeproject.com/Articles/8285/GAC-API-Interface

Reply:
This is likely of interest: https://github.com/dotnet/msbuild/blob/f2c4bfd563f559daca27ea1cd8ae40db24e2e7cd/src/Tasks/NativeMethods.cs#L344-L351

I don't think we can depend on `gacutil.exe` existing: the VS install may be what installs it. We could maybe express a dependency on "some .NET SDK" and use that, but it's probably easier to use the COM object.

> it's probably easier to use the COM object

😲

Reply:
I just spent a good chunk of time trying to get https://docs.microsoft.com/en-us/dotnet/api/system.enterpriseservices.internal.publish.gacremove?view=netframework-4.8 working to no avail (passing in file path, assembly name, pointing directly to dll, etc). MattCav suggested it was preferable to use some sort of API and keep the code simple so that's the route I'll go.

Reply:
Reusing `NativeMethods` was the way to go. The program does need to be running as admin for it to work, however.